src/compat/db_json.cpp

```diff
@@ -1168,6 +1361,339 @@ db_json_get_bool_as_str_from_value (const JSON_VALUE *doc)
   return db_json_bool_to_string (doc->GetBool());
 }
 
+static JSON_PATH_TYPE
+db_json_get_path_type (std::string &path_string)
+{
+  db_json_normalize_path (path_string);
+
+  if (path_string.empty())
+    {
+      return JSON_PATH_TYPE::JSON_PATH_EMPTY;
+    }
+  else if (path_string[0] == '$')
+    {
+      return JSON_PATH_TYPE::JSON_PATH_SQL_JSON;
+    }
+  else
+    {
+      return JSON_PATH_TYPE::JSON_PATH_POINTER;
+    }
+}
+
+/*
+* db_json_convert_pointer_to_sql_path ()
+* pointer_path (in)
+* doc (in)
+* path_has_effect(out) if path does not identify any value in the JSON document will be ignored.
+* sql_path_out (out): the result
+* A pointer path is converted to SQL standard path
+* Example: /0/name1/name2/2 -> $[0]."name1"."name2"[2]
+*/
+int
+db_json_convert_pointer_to_sql_path (const char *pointer_path, const JSON_DOC &doc,
+				     std::string &sql_path_out, bool &path_has_effect)
+{
+  std::string pointer_path_string (pointer_path);
+  JSON_PATH_TYPE json_path_type = db_json_get_path_type (pointer_path_string);
+
+  if (json_path_type == JSON_PATH_TYPE::JSON_PATH_EMPTY ||
+      json_path_type == JSON_PATH_TYPE::JSON_PATH_SQL_JSON)
+    {
+      // path is not JSON path format; consider it SQL path.
+      sql_path_out = pointer_path_string;
+      return NO_ERROR;
+    }
+
+  std::string result = "$";
+  std::string token;
+  unsigned int i;
+  unsigned int start_index;
+  unsigned int pointer_path_length = pointer_path_string.length();
+  std::unordered_map<std::string, std::string> special_chars;
+  special_chars.insert (std::make_pair ("~0", "~"));
+  special_chars.insert (std::make_pair ("~1", "/"));
+  special_chars.insert (std::make_pair ("%20", " "));
+  special_chars.insert (std::make_pair ("%E2%82%AC", "€"));
+
+  // starting the conversion of path
+  for (i = 0; i < pointer_path_length; ++i)
+    {
+      JSON_VALUE *resulting_json_parent;
+      JSON_POINTER p (pointer_path_string.substr (0, i).c_str());
+
+      if (pointer_path_string[i] == '/')
+	{
+	  i++;
+	  bool is_number = false;
+	  token.clear();
+
+	  // get current token between "/"
+	  // we need to know if it was a number or a string because we need to differentiate
+	  // between an index and an object name with the same token
+	  // for example "/0" can mean the element at index 0 from json_array or json_object with name "0"
+	  while (i < pointer_path_length && pointer_path_string[i] != '/')
+	    {
+	      if (pointer_path_string[i] >= '0' && pointer_path_string[i] <= '9' && !is_number)
+		{
+		  is_number = true;
+		}
+
+	      token += pointer_path_string[i++];
+	    }
+
+	  if (!p.IsValid())
+	    {
+	      return ER_JSON_INVALID_PATH;
+	    }
+
+	  resulting_json_parent = const_cast<JSON_VALUE *> (reinterpret_cast<const JSON_VALUE *> (p.Get (doc)));
```

Comment:
can't we remove these casts altogether?

Reply:
I can use only the const_cast without the reinterpret. but the cast is necessary because p.Get() method returns a const JSON_VALUE*